msm/__main__.py

```diff
@@ -116,18 +116,20 @@ def add_search_args(subparser, skill_is_optional=False):
         ),
         'info': lambda: skill_info(msm.find_skill(args.skill, args.author))
     }
-    try:
-        result = main_functions[args.action]()
-        if result is False:
-            return 1
-        if isinstance(result, str):
-            printer(result)
-        return 0
-    except MsmException as e:
-        exc_type = e.__class__.__name__
-        printer('{}: {}'.format(exc_type, str(e)))
-        return get_error_code(e.__class__)
-
+    with msm.lock:
+        try:
+            result = main_functions[args.action]()
+            if result is False:
+                return 1
+            if isinstance(result, str):
+                printer(result)
+            return 0
+        except MsmException as e:
+            exc_type = e.__class__.__name__
+            printer('{}: {}'.format(exc_type, str(e)))
+            return get_error_code(e.__class__)
+        finally:
+            msm.write_skills_data()
```

Comment:
Requiring this after any api call breaks the python api (used by the installer skill for example). I think we should either make it clear msm's python api shouldn't be used (and change the installer skill) or do something else like add a `@save_skills_data` decorator that functions are marked with in the `MycroftSkillsManager` class.

Reply:
Since I changed the Api i updated the minor version :) But you do have a point. (And I have made the corresponding changes in the core and installer skill.

I like the `@save_skills_data` decorator the case I don't really know how to handle is if the `MycroftSkillsManager.install` is passed to the apply-method. Is there a good way to work around that?

Basically strip the decorator if it exists when running the apply.

Reply:
I could maybe add an optional keyword parameter to the decorator that would disable the saving. In this case I could also handle the locking automatically...
Or set a flag on the function called or something.

Reply:
I'll give it a go. It would be really nice if this was automagic like this.

Reply:
Ok, done some testing and the cleanest way I've found is having the decorator adding an optional parameter and have the apply method to check for the decorator and send the disable parameter if it's there:

decorator:
```python
def save_skills_data(func):
    @wraps(func)
    def func_wrapper(self, *args, **kwargs):
        # Check for __msm_write keyword parameter
        do_write =  kwargs.get('__msm_write', True)
        # Clean the write parameter if exists
        if '__msm_write' in kwargs:
            kwargs.pop('__msm_write')
        try:
            ret = func(self, *args, **kwargs)
        finally:
            if do_write:
                self.write_skills_data()
        return ret

    func_wrapper.save_skills_data = True
    return func_wrapper
```

And the apply:

```python
def apply(self, func, skills):
        """Run a function on all skills in parallel"""

        def run_item(skill):
            try:
                if 'save_skills_data' in dir(func):
                    func(skill, __msm_write=False)
                else:
                    func(skill)
    [....]
```

Does this look sane?

Reply:
Hmm, no. yours is better since mine won't work if the msm.install/update/etc is wrapped in a function...